lib/features/build_server/build_provider.dart

```diff
@@ -0,0 +1,265 @@
+  /// Terminates the isolate which hosts the project with the id [projectId].
+  void kill(String projectId) {
+    final MainInterface? interface = _interfaces[projectId];
+
+    if (interface != null) {
+      interface.isolate.isolate.kill(priority: Isolate.immediate);
+    }
+
+    if (_availableProjects.containsKey(projectId)) {
+      _availableProjects.remove(projectId);
+    }
+  }
+
+  /// Returns a list of all projects which have an active connection.
+  List<String> get projectIds {
+    return _availableProjects.keys.toList();
+  }
+
+  /// Returns the name of the project with the [projectId].
+  ///
+  /// If the project is not available, an argument error is thrown.
+  String projectName(String projectId) {
+    final String? name = _availableProjects[projectId];
+
+    if (name != null) {
+      return name;
+    }
+
+    throw ArgumentError('Project with id $projectId does not exists');
+  }
+
+  /// Handles a new websocket message.
+  Future<void> _handleWebSocketMessage(ServerSession session, dynamic message) async {
+    if (message is ProjectInitRequest) {
+      await _handleProjectInit(session, message);
+      return;
+    }
+
+    final String id = _ensureId(session);
+
+    if (message == reserveBuildRequest) {
+      _projectCapacityQueues[id]!.addLast(session);
+
+      if (_isFirstInQueue(session, id)) {
+        await _activateSession(session, id);
+      }
+    }
+
+    if (message == unsubscribeRequest) {
+      await _activateNextSession(id, removing: session);
+      return;
+    }
+
+    if (!_isFirstInQueue(session, id)) {
+      session.send(isInactiveSessionResponse);
+      return;
+    }
+
+    if (message is ProjectUpdateRequest) {
+      await _handleProjectUpdate(session, message, id);
+    }
+
+    if (message is FileUpdate) {
+      await _handleFileUpdate(session, message, id);
+    }
+
+    if (message == projectBuildRequest) {
+      await _getInterface(id).build();
+    }
+
+    if (message == getBinaryRequest) {
+      await _handleBinaryRequest(session, message, id);
+    }
+  }
+
+  /// Checks if the [session] is the first in the queue for
+  /// the project with the id [id].
+  bool _isFirstInQueue(ServerSession session, String id) {
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      return identical(_projectCapacityQueues[id]?.first, session);
+    }
+
+    return false;
+  }
+
+  /// Activates the first session in the queue for the project with the id [id].
+  ///
+  /// If [removing] is not null, the session is removed from the queue.
+  Future<void> _activateNextSession(String id, {ServerSession? removing}) async {
+    if (removing != null) {
+      _projectCapacityQueues[id]?.remove(removing);
+    }
+
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      _activateSession(_projectCapacityQueues[id]!.first, id);
+    }
+  }
+
+  /// Activates the [session] for the project with the id [id].
+  Future<void> _activateSession(ServerSession session, String id) async {
```

Comment:
Did you forget to set `session.storage['id']` here? If you activate another session it would not contain the `id` in its `storage` from my understanding, as you simply return before assigning it in `_handleProjectInit`. If the second client would disconnect `wsOnDone` should be unable to activate another session.

Reply:
Wrong function but the problem is real. Fixed it by ensuring each session gets an ID assigned when they send project init request.

Reply:
👍🏻 Yeah, thats why I asked how the init request worked, because I couldn’t determine which function should be responsible :)